crates/ra_ide_api/src/name_ref_kind.rs

```diff
@@ -0,0 +1,99 @@
+use ra_db::FileId;
+use ra_syntax::{AstNode, AstPtr, ast};
+use hir::Either;
+use crate::db::RootDatabase;
+use test_utils::tested_by;
+
+pub enum NameRefKind {
+    Method(hir::Function),
+    Macro(hir::MacroByExampleDef),
+    FieldAccess(hir::StructField),
+    AssocItem(hir::ImplItem),
+    Def(hir::ModuleDef),
+    SelfType(hir::Ty),
+    Pat(AstPtr<ast::Pat>),
+    SelfParam(AstPtr<ast::SelfParam>),
+    GenericParam(u32),
+}
+
+pub(crate) fn classify_name_ref(
+    db: &RootDatabase,
+    file_id: FileId,
+    name_ref: &ast::NameRef,
+) -> Option<NameRefKind> {
+    use NameRefKind::*;
+    let analyzer = hir::SourceAnalyzer::new(db, file_id, name_ref.syntax(), None);
```

Comment:
Can `SourceAnalyzer`s be reused somehow?

Reply:
Good question! In general, `SourceAnalyzer` should be reused on per-function basis. 

Let's pass it in as an argument to the function? If the caller needs to classify several refs, they can then re-use the same analyzer.

Reply:
I was confused because `SourceAnalyzer::new` takes (and uses) the `&SyntaxNode` of as an argument.

Reply:
Yeah, but the Node needn't be exactly the one you will be asking questions about. Any child node of the function would do.

The API is slightly confusing (nothing prevents you from passing a node from *another* function, and that would be an error), but it is in general not clear yet what is the best way to bridge semantic info from hir and source-level info from parse trees. `SourceAnalyzer` is definitely better than the thing we had before :-)